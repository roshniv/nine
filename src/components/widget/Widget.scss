.widget {
  //mobile
  padding: 33px 30px 30px 33px;
  background-color: white;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
  border-radius: 4px;

  &__title {
    margin: 0;
    font-weight: bold;
    font-size: 18px;
  }
  &__content {
    display: flex;
    margin-top: 36px;
  }
  &__loading{
    display: flex;
  }
  &__weather-img {
    width: 40%;
    padding-right: 10px;
    object-fit: cover;
  }
  &__weather-info {
    padding-left: 10px;
  }
  &__temperature {
    font-size: 42px;
    font-weight: bold;
  }
  &__wind {
    min-height: 20px;
  }
  &__wind-status {
    font-size: 12px;
    font-weight: bold;
    margin-right: 6px;
  }
  &__wind-info {
    font-size: 12px;
  }
}
