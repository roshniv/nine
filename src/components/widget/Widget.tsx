import React, { useEffect, useState } from 'react';
import './Widget.scss';
import { MoonLoader } from 'react-spinners';
import axios from 'axios';
import Day from './icons/day.svg';
import Rainy from './icons/rainy.svg';
import Snowy from './icons/snowy.svg';

interface IWeather {
    name: string,
    main: {
        temp: number
    },
    wind: {
        speed: number,
        deg: number
    },
    weather: {
        main: string
    }[]
}

interface IWeatherProps {
    isCelsius: boolean,
    showWind: boolean
}

const geoOptions = {
  enableHighAccuracy: true,
  timeout: 10000,
  maximumAge: 0,
};

const Widget = ({ isCelsius, showWind }: IWeatherProps): React.ReactElement => {
  const [weatherData, setWeatherData] = useState({} as IWeather);
  const [error, setError] = useState(false as boolean);
  const [loading, setLoading] = useState(true as boolean);

  // fetch weather and geo data when load component
  useEffect(() => {
    navigator.geolocation.getCurrentPosition((pos) => {
      const { latitude, longitude } = pos.coords;
      axios.get(`https://api.openweathermap.org/data/2.5/weather?lat=${latitude}&lon=${longitude}&appid=${'ee7405469955e124ef77214a739627c5'}`)
        .then((res) => {
          setWeatherData(res.data);
          setLoading(false);
        })
        .catch(() => {
          setError(true);
          setLoading(false);
        });
    }, () => {
      setError(true);
      setLoading(false);
    }, geoOptions);
  }, []);

  const calculateTemperature = () => {
    const { temp } = weatherData.main;
    const f = Math.round((temp - 273.15) * 9 / 5 + 32);
    const c = Math.round(temp - 273.15);
    return isCelsius ? c : f;
  };

  const weatherImageSrc = () => {
    let imgSrc;
    switch (weatherData.weather[0].main) {
      case 'Rain':
        imgSrc = Rainy;
        break;
      case 'Snow':
        imgSrc = Snowy;
        break;
      default:
        imgSrc = Day;
    }
    return imgSrc;
  };

  const renderWind = () => {
    if (showWind) {
      const windDegVal = Math.round((weatherData.wind.deg / 22.5) + 0.5);
      const directions = ['N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE', 'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW'];
      const direction = directions[(windDegVal % 16)];
      const windSpeed = Math.round(weatherData.wind.speed);
      return (
        <>
          <span className="widget__wind-status">Wind</span>
          <span className="widget__wind-info">
            {direction}
            {' '}
            {windSpeed}
            km/h
          </span>
        </>
      );
    }
    return null;
  };

  const renderContent = () => {
    if (loading) {
      return (
        <div className="widget__loading" data-testid="loading">
          <MoonLoader color="black" loading={loading} size={48} />
        </div>
      );
    } if (error) {
      return (
        <>
          <p>Please try later</p>
        </>
      );
    }
    return (
      <>
        <img src={weatherImageSrc()} aria-label="weather-icon" alt="weather" className="widget__weather-img" />
        <div className="widget__weather-info" data-testid="widget-info">
          <div className="widget__city">{weatherData.name}</div>
          <div className="widget__temperature">
            {calculateTemperature()}
            &#176;
          </div>
          <div className="widget__wind" data-testid="widget-wind">
            {renderWind()}
          </div>
        </div>
      </>
    );
  };

  return (
    <div className="widget">
      <div className="widget__title" data-testid="widget-title">TITLE</div>
      <div className="widget__content">
        {renderContent()}
      </div>
    </div>
  );
};

export default Widget;
