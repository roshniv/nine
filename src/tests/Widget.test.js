import React from 'react';
import { cleanup, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Widget from '../components/widget/Widget';
import {
  mockGetlocationSuccess,
  mockGetlocationError,
  mockAxiosGetSuccess,
  sydneyMockWeatherData,
  mockAxiosGetError,
} from './utils/mocks';

afterEach(cleanup);

jest.mock('axios');

// get html fragment for snapshot testing
let asFragment;

describe('Widget component', () => {
  const initialWidgetTitle = 'title';

  it('should render weather information if user allow to get location in browser', async () => {
    mockAxiosGetSuccess(sydneyMockWeatherData);
    mockGetlocationSuccess();

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(initialWidgetTitle.toUpperCase())).toBeInTheDocument();
    expect(screen.getByTestId('widget-info')).toBeInTheDocument();
  });

  it('show loading when component start loading', () => {
    const rendered = render(<Widget widgetTitle="title" isCelsius showWind />);
    asFragment = rendered.asFragment;

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('loading')).toBeInTheDocument();
  });

  it('renders error message if cannot get position', async () => {
    mockGetlocationError();

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Please try later', { selector: 'p' })).toBeInTheDocument();
  });

  it('renders error message if cannot get weather information', async () => {
    mockGetlocationSuccess();
    mockAxiosGetError();

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Please try later', { selector: 'p' })).toBeInTheDocument();
  });

  it('shows a sunny icon if it is a good day!', async () => {
    mockGetlocationSuccess();
    mockAxiosGetSuccess(sydneyMockWeatherData);

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('img', { name: 'weather-icon' })).toHaveAttribute('src', 'day.svg');
  });

  it('shows a rainy icon if it is a rainy day!', async () => {
    const mockRainyData = { ...sydneyMockWeatherData };
    mockRainyData.weather[0].main = 'Rain';

    mockGetlocationSuccess();
    mockAxiosGetSuccess(mockRainyData);

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('img', { name: 'weather-icon' })).toHaveAttribute('src', 'rainy.svg');
  });

  it('shows a snowy icon if it is a snowy day!', async () => {
    const mockSnowyData = { ...sydneyMockWeatherData };
    mockSnowyData.weather[0].main = 'Snow';

    mockGetlocationSuccess();
    mockAxiosGetSuccess(sydneyMockWeatherData);

    await act(async () => {
      const rendered = render(<Widget widgetTitle={initialWidgetTitle} isCelsius showWind />);
      asFragment = rendered.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('img', { name: 'weather-icon' })).toHaveAttribute('src', 'snowy.svg');
  });
});
