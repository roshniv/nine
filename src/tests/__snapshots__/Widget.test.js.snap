// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget component renders error message if cannot get position 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <p>
        Please try later
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component renders error message if cannot get weather information 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <p>
        Please try later
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component should render weather information if user allow to get location in browser 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <img
        alt="weather"
        aria-label="weather-icon"
        class="widget__weather-img"
        src="day.svg"
      />
      <div
        class="widget__weather-info"
        data-testid="widget-info"
      >
        <div
          class="widget__city"
        >
          Sydney
        </div>
        <div
          class="widget__temperature"
        >
          12째
        </div>
        <div
          class="widget__wind"
          data-testid="widget-wind"
        >
          <span
            class="widget__wind-status"
          >
            Wind
          </span>
          <span
            class="widget__wind-info"
          >
            NW 1km/h
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component show loading when component start loading 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <div
        class="widget__loading"
        data-testid="loading"
      >
        <span
          class="css-18jvrwa"
        >
          <span
            class="css-19yb4yo"
          />
          <span
            class="css-1ad1jmu"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component shows a rainy icon if it is a rainy day! 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <img
        alt="weather"
        aria-label="weather-icon"
        class="widget__weather-img"
        src="rainy.svg"
      />
      <div
        class="widget__weather-info"
        data-testid="widget-info"
      >
        <div
          class="widget__city"
        >
          Sydney
        </div>
        <div
          class="widget__temperature"
        >
          12째
        </div>
        <div
          class="widget__wind"
          data-testid="widget-wind"
        >
          <span
            class="widget__wind-status"
          >
            Wind
          </span>
          <span
            class="widget__wind-info"
          >
            NW 1km/h
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component shows a snowy icon if it is a snowy day! 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <img
        alt="weather"
        aria-label="weather-icon"
        class="widget__weather-img"
        src="snowy.svg"
      />
      <div
        class="widget__weather-info"
        data-testid="widget-info"
      >
        <div
          class="widget__city"
        >
          Sydney
        </div>
        <div
          class="widget__temperature"
        >
          12째
        </div>
        <div
          class="widget__wind"
          data-testid="widget-wind"
        >
          <span
            class="widget__wind-status"
          >
            Wind
          </span>
          <span
            class="widget__wind-info"
          >
            NW 1km/h
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Widget component shows a sunny icon if it is a good day! 1`] = `
<DocumentFragment>
  <div
    class="widget"
  >
    <div
      class="widget__title"
      data-testid="widget-title"
    >
      TITLE
    </div>
    <div
      class="widget__content"
    >
      <img
        alt="weather"
        aria-label="weather-icon"
        class="widget__weather-img"
        src="day.svg"
      />
      <div
        class="widget__weather-info"
        data-testid="widget-info"
      >
        <div
          class="widget__city"
        >
          Sydney
        </div>
        <div
          class="widget__temperature"
        >
          12째
        </div>
        <div
          class="widget__wind"
          data-testid="widget-wind"
        >
          <span
            class="widget__wind-status"
          >
            Wind
          </span>
          <span
            class="widget__wind-info"
          >
            NW 1km/h
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
